File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    r,i,idx=[],0,27
    dic=dict()
    for i in range(26):
        dic[chr(i+65)]=i+1
    i=0
    while i<len(msg):
        t,cnt="",0
        while i+cnt<len(msg) and t+msg[i+cnt] in dic:
            t+=msg[i+cnt]
            cnt+=1
        r+=[dic.get(t)]
        if i+cnt<len(msg):
            dic[t+msg[i+cnt]]=idx
            idx+=1
        i+=cnt
    return r

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

vector<int> solution(string msg) {
    vector<int> r;
    map<string,int> dic;
    string t;
    int cnt,i=0,idx=27;
    for(int i=0;i<26;i++) {
        t=(char)i+65;
        dic[t]=i+1;
    }
    while(i<msg.size()) {
        t="",cnt=0; 
        while(i+cnt<msg.size() and dic.find(t+msg[i+cnt])!=dic.end()) {    
            t+=msg[i+cnt++];
        }
        r.push_back(dic[t]);
        if(i+cnt<msg.size()) {
            dic[t+msg[i+cnt]]=idx++;
        }
        i+=cnt;
    }
    return r;
}